### PR TITLE
Implement returning for postgresql

### PR DESCRIPTION
**Pull request by 00dav00**
### Summary

This is an attempt to implement #34237

It adds a class method that will receive the attributes that should be returned on record creation or update because they are calculated inside postgresql.

### Steps to reproduce
 
**1. Create a model**
 
```Ruby
class Order < ApplicationRecord
end
```
 
**2. Create a migration with a DB function as `default`.**
 
```Ruby
class CreateOrders < ActiveRecord::Migration[5.2]
  def change
    create_table :orders do |t|
      t.string :uuid, default: -> { "gen_random_uuid()" }
      t.references :user
      t.references :product
      t.text :comment
 
      t.timestamps
    end
  end
end
```
 
`rails db:migrate`
 
**3. Create a new order**
 
`rails console`
 
```
order = Order.create(user_id: 1, product_id: 1)
```
 
### Desired behavior
I would like `create` to set `uuid` (by the postgres default) **and** for it to set the new `uuid` value into the `order` AR instance, as it does with the `id`.
 
```
> order
# =>
# +----+--------------------------------------+---------+------------+---------+----------------+----------------+
# | id | uuid                                 | user_id | product_id | comment | created_at     | updated_at     |
# +----+--------------------------------------+---------+------------+---------+----------------+----------------+
# | 1  | bec94fe0-f7c3-4ede-8b25-c4bce702ec00 | 1       | 1          |         | 2018-10-17 ... | 2018-10-17 ... |
# +----+--------------------------------------+---------+------------+---------+----------------+----------------+
```
 
### Actual behavior
The `uuid` does get set in the db, but it is not returned into the `order` instance.

### CODE REVIEW

**Comment by natematykiewicz** [comment]
My company's Rails app (using Postgres) has quite a few columns that are managed by database triggers. A `RETURNING` statement to populate them on insert/update sounds amazing!

**Comment by davegson** [comment]
just a gentle fyi - the failing checks should probably be tackled

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb [hunk 1]
```diff
@@ -162,10 +162,10 @@ def exec_insert_all(sql, name) # :nodoc:
       #
       # If the next id was calculated in advance (as in Oracle), it should be
       # passed in as +id_value+.
-      def insert(arel, name = nil, pk = nil, id_value = nil, sequence_name = nil, binds = [])
+      def insert(arel, name = nil, pk = nil, id_value = nil, sequence_name = nil, binds = [], returning = [])
         sql, binds = to_sql_and_binds(arel, binds)
-        value = exec_insert(sql, name, binds, pk, sequence_name)
-        id_value || last_inserted_id(value)
+        value = exec_insert(sql, name, binds, pk, sequence_name, returning)
+        { 'id' => id_value }.merge(last_inserted_values(value))
```
Comment (natematykiewicz) [inline activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:168]:
```suggestion
        { "id" => id_value }.merge(last_inserted_values(value))
```

File: activerecord/lib/active_record/persistence.rb [hunk 2]
```diff
@@ -929,11 +930,12 @@ def _update_record(attribute_names = self.attribute_names)
     def _create_record(attribute_names = self.attribute_names)
       attribute_names = attributes_for_create(attribute_names)
 
-      new_id = self.class._insert_record(
+      new_values = self.class._insert_record(
         attributes_with_values(attribute_names)
       )
 
-      self.id ||= new_id if @primary_key
+      self.id ||= new_values['id'] if @primary_key
```
Comment (natematykiewicz) [inline activerecord/lib/active_record/persistence.rb:937]:
```suggestion
      self.id ||= new_values["id"] if @primary_key
```

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb [hunk 3]
```diff
@@ -545,6 +545,12 @@ def single_value_from_rows(rows)
           row && row.first
         end
 
+        def last_inserted_values(result)
+          return {} if result.rows.empty?
+
+          Hash[result.columns.zip(result.rows.first).map {|key, value| [key, value]}]
```
Comment (natematykiewicz) [inline activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:551]:
```suggestion
          Hash[result.columns.zip(result.rows.first).map { |key, value| [key, value] }]
```

File: activerecord/lib/active_record/persistence.rb [hunk 4]
```diff
@@ -929,11 +930,12 @@ def _update_record(attribute_names = self.attribute_names)
...
+      self.assign_attributes(new_values.except('id'))
```
Comment (natematykiewicz) [inline activerecord/lib/active_record/persistence.rb:938]:
```suggestion
      self.assign_attributes(new_values.except("id"))
```

**Review by natematykiewicz — CHANGES_REQUESTED** [review]

**Comment by natematykiewicz** [comment]
I just "requested changes" to make Rubocop happy.

**Comment by gmaliar** [comment]
@00dav00 this is amazing! My company would appreciate this PR a lot, is there anyway I can help? Is it okay if I try to fix the breaking tests?

**Comment by 00dav00** [comment]
Hi @gmaliar , the biggest problem this PR has is that the feature is not officially approved, that's why I haven't continue working on it. Here is the [rails forum issue](https://discuss.rubyonrails.org/t/feature-request-implement-returning-for-postgresql-calculated-columns/76058), if you know someone that can get admins attention on this it would be greatly appreciated.

**Comment by jrochkind** [comment]
@matthewd suggested in #34237 that they thought some functionality in this area did make sense as part of AR core, perhaps they could provide some feedback or advice on this PR to try to move it along. 

Specifically they said:

> To my eye, the ideal behaviour would be @tbuehlmann's explicit self.returning_attributes = (for manual additions reflecting things like triggers), plus automatic detection of unparseable default expressions [to always be returned on insert] and computed columns [to always be returned on insert or update].

**Comment by jorge-precich** [comment]
Loving the idea and the PR. My company is also having issues with `gen_random_uuid()` from PostgreSQL. You should really follow @jrochkind suggestion. I'd like to help with any you may need 👍

**Comment by 00dav00** [comment]
Hi  @jorge-precich , I do started this PR based on that issue that @jrochkind mentioned and would definitely love any feedback  or advice from @matthewd . I guess that maybe if the [rails forrum issue](https://discuss.rubyonrails.org/t/feature-request-implement-returning-for-postgresql-calculated-columns/76058) gets more interaction the rails core team may be open to consider the addition.

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb [hunk 5]
```diff
@@ -131,8 +131,8 @@ def exec_query(sql, name = "SQL", binds = [], prepare: false)
       # Executes insert +sql+ statement in the context of this connection using
       # +binds+ as the bind substitutes. +name+ is logged along with
       # the executed +sql+ statement.
-      def exec_insert(sql, name = nil, binds = [], pk = nil, sequence_name = nil)
-        sql, binds = sql_for_insert(sql, pk, binds)
+      def exec_insert(sql, name = nil, binds = [], pk = nil, sequence_name = nil, returning = [])
```
Comment (bf4) [inline activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:134]: given that returning is optional, I suspect the extra allocation here is undesired

as an aside, I believe returning would be relevant to handling composite primary keys

https://github.com/composite-primary-keys/composite_primary_keys/blob/c5c27220e844d95e568728dd5fde705e83adf814/lib/composite_primary_keys/connection_adapters/postgresql/database_statements.rb#L12-L18

Comment (00dav00) [reply activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:134]: Hi @bf4 , the reason returning is sent here is for the values to make it's way into the [place where the sql statement is build](https://github.com/rails/rails/pull/39968/files/543627f47cbb69d0007e7f43c94372c18f2d4d4d#diff-e6fd03cad5e3437c76b6c5db106359124dc9feab8341ade39b9ae54af001fac9R86)

Comment (bf4) [reply activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:134]: I meant allocating the array when its use is the exception in what I guess is a hotspot is worth considering the non-nil default

Comment (natematykiewicz) [reply activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:134]: Yeah, the suggestion would be to do `returning = nil`, and then later in spots where you expect it to be an array, account for it to be `nil` as well.

This means one less array for queries that don't use the returning feature, and that extra array can add up for a function that's called a lot in a web request.

Comment (00dav00) [reply activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:134]: I was following this logic:
- String default: nil (like `pk`)
- Array default: empty array (like `bind`)

But I see what you mean @bf4 @natematykiewicz and will do the change. Thanks for the suggestion.

**Comment by natematykiewicz** [comment]
I still would really like this to get merged

**Comment by jrochkind** [comment]
I am also very interested in this functionality. We have a PR. Is there anything we can do to help? Is there any way to get Rails committer attention?

**Comment by 00dav00** [comment]
@natematykiewicz @jrochkind I will rebase against the last changes in master to prevent it from being closed. I'm not sure what we can do get the attention of a committer. Do you have any suggestions?

**Comment by gmaliar** [comment]
@00dav00 perhaps we should turn this into a gem instead?

**Comment by natematykiewicz** [comment]
@00dav00 does this run on update as well? In order to get approved, it'll probably need to run on update too, so that it's done on "save".

**Comment by bf4** [comment]
@matthewd Does this meet your expectations for a feature PR per https://github.com/rails/rails/issues/34237#issuecomment-522204919

>>> I tackled that exact problem in a [POC over here](https://github.com/tbuehlmann/active_record-returning_attributes). It wasn't that simple as I had to override quite some AR core code. So it's not a very great implementation and will probably break as soon as internals change.

>> is this functionality that would make sense as a part of ActiveRecord core ?

> Yes. To my eye, the ideal behaviour would be @tbuehlmann's explicit self.returning_attributes = (for manual additions reflecting things like triggers), plus automatic detection of unparseable default expressions [to always be returned on insert] and computed columns [to always be returned on insert or update].

**Comment by natematykiewicz** [comment]
The composite primary keys (CPK) gem is half of my use case actually.

We use it at my work, and it would remove a lot of that gem's monkey patches. The monkey patching necessary means that the CPK gem usually breaks with every minor version of active record. I would really recommend against making this it's own gem as you'll constantly be chasing your tail patching (just like CPK does).

My hope is I can use CPK gem and this together.

```ruby
class Foo
  # Composite Primary Keys gem
  self.primary_keys = :a_id, :b_id
  
  # Populated via triggers on insert/update
  returning :c, :d
end
```

My expectation then would be that the `RETURNING` clause would be `a_id, b_id, c, d` on `INSERT` and `UPDATE`, so that we could have the IDs but also the 2 extra columns that are populated via triggers.

It would be amazing if this works.

There's no tests for an update currently, so I'm not so sure if it works for an update.

For my case, I'd need to make sure that `c` and `d` were returned, or else I'll still need to reload to get their values. Having `a_id` and `b_id` returned on update is unnecessary, but doesn't hurt anything so I would expect them to be returned on `UPDATE` as well.

**Comment by ibrahima** [comment]
I was researching generating default values in the database and stumbled across this issue. If it helps, here's a really simple test case I wrote to demonstrate the issue: https://github.com/ibrahima/rails_experiments/pull/1/files

I can try and use this PR with that test case to validate that it solves the problem. I've glanced through the PR and it looks pretty reasonable to me, but as with everyone else, I am not a Rails committer, so I don't know how to get this looked at. I think it'd be great to have this in ActiveRecord though! It's great to be able to use DB functionality where possible.

A workaround I've seen mentioned is to add an `after_create { reload }` hook but that feels kludgey to me.

**Comment by md5** [comment]
While this may not be a blocker for getting this merged, this PR doesn't touch any code in `lib/arel`, which IMHO makes it harder to maintain and extend going forward. The same issue applies to how `#insert_all` was implemented directly in `ActiveRecord::Persistence` without touching Arel. It seems like in both of these cases the `RETURNING` functionality should have representation under `Arel::Nodes` and the SQL should be generated using code in `Arel::Visitors`, though I suppose the same goes for the original implementation of `RETURNING id`, which also does not live in Arel.

**Comment by 00dav00** [comment]
You are correct @md5 , I based my implementation in `insert_all`. I see where you are coming from for this suggestion, nevertheless not sure making the change will do anything to move this forward  XD

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb [hunk 6]
```diff
@@ -560,6 +568,12 @@ def single_value_from_rows(rows)
           row && row.first
         end
 
+        def last_inserted_values(result)
+          return {} if result.rows.empty?
+
+          Hash[result.columns.zip(result.rows.first).map { |key, value| [key, value] }]
```
Comment (bf4) [inline activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:574]: Fwiw this is inefficient from an object allocation pov.  It's probably not great in a hotspot

File: activerecord/lib/active_record/persistence.rb [hunk 7]
```diff
@@ -971,11 +972,16 @@ def _update_record(attribute_names = self.attribute_names)
...
 
-      new_id = self.class._insert_record(
+      result = self.class._insert_record(
         attributes_with_values(attribute_names)
       )
 
-      self.id ||= new_id if @primary_key
+      if result.is_a?(Hash)
+        self.id ||= result["id"] if @primary_key
+        self.assign_attributes(result.except("id"))
```
Comment (bf4) [inline activerecord/lib/active_record/persistence.rb:981]: Why do in two lines what looks like only needs one?

File: activerecord/test/cases/returning_test.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  fixtures :books
+
+  # def setup
+  #   Arel::Table.engine = nil # should not rely on the global Arel::Table.engine
+  # end
+
+  # def teardown
+  #   Arel::Table.engine = ActiveRecord::Base
+  # end
+
```
Comment (bf4) [inline activerecord/test/cases/returning_test.rb:16]: I too would would write fewer tests without Rails core attention one way or another:)

**Comment by zzak** [comment]
Trying to take a quick stab at summarizing this PR:

* :eyes: Need a review on this implementation (similar to `insert_all`)
* 📈 Current implementation could _maybe_ be optimized further
* It seems that test coverage could improve

That said, I think the first point is by far the most important.

We can also add more tests and refactor after that.

File: activerecord/lib/active_record/persistence.rb (hunk 7 above)
Comment (00dav00) [reply activerecord/lib/active_record/persistence.rb:981]: I can check if keeping only the `assign_attributes` call can work, wanted to stay close to current implementation

**Comment by 00dav00** [comment]
Hi @zzak  thanks so much for checking this!
Is there something I have to/can do about the `Need a review on this implementation` part or is this an rails team task? Please let me know :)

**Comment by zzak** [comment]
@00dav00 For review, it helps to have someone with additional context and in this particular case my best guess is @matthewd -- so we may just have to wait for them on this.

**Comment by 00dav00** [comment]
Understood @zzak , thanks for clearing this up. I guess nothing to do but wait for this reach @matthewd 's radar.

**Comment by natematykiewicz** [comment]
Kasper is doing some PR review this week. I suggested reviewing this PR.

He has some feedback:
https://twitter.com/natematykiewicz/status/1421342172796882944

**Comment by jrochkind** [comment]
Thanks @natematykiewicz . The feedback "try closing it and then reopening a new PR doing it in a different way" -- without having actually reviewed the code here, and thus with absolutely no tips on what sort of different way might be good -- seems to me profoundly discouraging and unhelpful.  I don't know why anyone would put their time into another PR based on that, I sure wouldn't. 

My conclusion is basically just that nobody on Rails core team is interested in this, and there is no way to get a PR reviewed in that case, so it's unlikely to be a good use of time to try an alternate PR. 

I wonder if there's any way to get this feature with a relatively light monkey-patch that seems possibly likely to survive future releases? That would seem to be the only hope.

**Comment by 00dav00** [comment]
@natematykiewicz Thanks for checking with Kasper, I think I will just let this be closed by GH bots cause it looks like there is no way of getting this reviewed.

@jrochkind I'm not sure about monkey patching for this case may because it may be to fragile given that the changes sit at the core of active record.

**Comment by natematykiewicz** [comment]
Monkey patching it is a really bad idea. The Composite Primary Keys gem does a similar monkey patch to return your composite primary key columns back, and the gem is very unstable between patch versions of ActiveRecord. Having something like this in Rails would be a huge help to a lot of things. I see 2 things that I'm guessing would help this PR along:

1. It's currently Postgres only
2. It uses the word `returning`, which is Postgres term

I guess a big question is, does MySQL have a similar feature available that this could also support? Basecamp uses MySQL, so making this be a feature for that would probably help this gain traction.

**Comment by jrochkind** [comment]
Googling, it's unfortuantely looking to me like MySQL offers no equivalent. :(

**Comment by deepj** [comment]
This feature is supported by PostgreSQL and SQLite (more limited than in PostgreSQL). MySQL doesn't support it yet.

- https://www.postgresql.org/docs/12/dml-returning.html
- https://www.sqlite.org/lang_returning.html

**Comment by jrochkind** [comment]
Here's a PR that just get merged for something supported by Postgres but not MySQL.  It may provide a model of how to structure code in Rails for features  only supported by some adapters and not others (as well as evidence that such features can be merged sometimes!). 

https://github.com/rails/rails/pull/41487

**Comment by jrochkind** [comment]
Here's another PR for `returning` on SQLite. https://github.com/rails/rails/pull/42955

Maybe if they can be reconciled to use simlar code, and the attention of any rails committers can be obtained (that's the hard part), there'd be a chance of merge?

And here's a merged PR for an unrelated feature that supports only postgres and SQLite, demonstrating that Rails is in some cases willing to merge AR features only supported on postgres and sqlite, not mysql. https://github.com/rails/rails/pull/40491

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb (hunk 6 above)
Comment (semaperepelitsa) [reply activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:574]: Simply `result.first` will work (from ActiveRecord::Result). It is already optimized as much as possible.

**Comment by OuYangJinTing** [comment]
> Here's another PR for `returning` on SQLite. #42955
> 
> Maybe if they can be reconciled to use similar code, and the attention of any rails committers can be obtained (that's the hard part), there'd be a chance of merge?
> 
> And here's yet another merged PR for an unrelated feature that supports only postgres and SQLite, demonstrating that Rails is in some cases willing to merge AR features only supported on postgres and sqlite, not mysql. #40491

Actually. MariaDB support returning clause too, then MariaDB and MySql use common adapter. So I think returing function should be more widely used, not just limited to postgresql. I have also proposed a feature in rails form: https://discuss.rubyonrails.org/t/feature-proposal-activerecord-update-all-delete-all-support-use-sqls-returning-clause-statements/78729

**Comment by 00dav00** [comment]
Hi @OuYangJinTing , I think this has more to do with getting the right person to review and agree on the changes than been able to make the feature broader, as @zzak commented before.

**Comment by fatkodima** [comment]
I think, this PR should be also extended to automatically support `virtual` columns, because currently there is a need to do a reload to get the value.

File: activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb [hunk 9]
```diff
@@ -160,12 +160,20 @@ def explain(arel, binds = []) # :nodoc:
       # which case the database will attempt to calculate the last inserted
       # id and return that value.
       #
+      # If "returning" param is provided in Postgresql, the method will return
+      # a hash with the values including the id
```
Comment (bf4) [inline activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:164]: changing the return type is, imho, problematic

**Comment by winnieoursbrun** [comment]
Would be great too for updates, specially with `increment_counter` method, with that we can have atomic counter for unique name generation based on a count, for example.

**Comment by JuPlutonic** [comment]
@winnieoursbrun FYI I used this psql code in Rails production for the same reason:

```ruby
  # foobars_id is a field used in unique name generation, e.g. with concatinated version of foobars
  def generate_number
    self.version_number = 1
    query = "select setval('foobars_id_seq'::regclass, nextval('foobars_id_seq'::regclass), false)"
    doc_num = ActiveRecord::Base.connection.select_values(query)[0]
    self.document.id = doc_num
    self.number = "#{doc_num.to_s.rjust(5, '0')}-#{self.version_number}"
end
```

**Comment by quentindemetz** [comment]
My company is using https://github.com/pennylane-hq/active_record-returning_attributes/tree/qdm/rails7 for the moment

**Comment by jrochkind** [comment]
It looks like this has finally been handled, seperately by separate development, here?  https://github.com/rails/rails/pull/39968

**Comment by jrochkind** [comment]
It looks like this has finally been handled, separately by separate development, here?  https://github.com/rails/rails/pull/48241

**Pull request closed without merging**